### ISSUE TITLE
`stack build; stack test` rebuilds everything twice

### ISSUE THREAD

**Issue by minad** [commercialhaskell/stack#4977]
### Steps to reproduce

1. `stack build`
2. `stack test`

I am using stackage nightly. The library of my project is rebuild by `stack test`.

### Expected

The library should be compiled once and the compiled and cached output should be used.

### Actual

`stack test` recompiles the library.

### Stack version

```
2.1.1
```

### Method of installation

* Official binary, downloaded from stackage.org or fpcomplete's package repository

**Comment by minad** [comment]
Maybe it has something to do with my Setup.hs? I try to create a repro if possible. If I recall correctly, the issue started to occur after I upgraded to stack version 2.

**Comment by gromakovsky** [comment]
I sometimes observed this behavior, but it was not deterministic and I never tried to investigate it. I think I didn't observe it during last few months though.

**Comment by sjakobi** [comment]
I can observe this reliably. Switching between `stack test` and `stack bench` rebuilds the library too. This issue is a big time waster in my experience.

**Comment by mihaimaruseac** [comment]
If the compile flags need to change between building test and just building lib + exe then Stack would have to recompile everything, just like any other build system (just closed a similar issue with Bazel on another repo)

If you can provide a minimal example, we can investigate to see if Stack is too eager in invalidating the build cache

**Comment by sjakobi** [comment]
Here's an example using https://github.com/nikita-volkov/vector-builder:

```
$ cd vector-builder
$ stack clean --full
$ stack build
vector-builder> configure (lib)
Configuring vector-builder-0.3.7.2...
vector-builder> build (lib)
Preprocessing library for vector-builder-0.3.7.2..
Building library for vector-builder-0.3.7.2..
[1 of 8] Compiling VectorBuilder.Prelude
[2 of 8] Compiling VectorBuilder.Core.Update
[3 of 8] Compiling VectorBuilder.Core.Builder
[4 of 8] Compiling VectorBuilder.MVector
[5 of 8] Compiling VectorBuilder.Builder
[6 of 8] Compiling VectorBuilder.Vector
[7 of 8] Compiling VectorBuilder.MonadPlus
[8 of 8] Compiling VectorBuilder.Alternative
vector-builder> copy/register
Installing library in /home/simon/src/vector-builder/.stack-work/install/x86_64-linux/da6ab40722cf26def321f3afbad12b99af2f327f84d5d7d92f8138157548da6d/8.6.5/lib/x86_64-linux-ghc-8.6.5/vector-builder-0.3.7.2-JJTkYJEXEoEG9xl9pXropy
Registering library for vector-builder-0.3.7.2..
$ stack test --no-run-tests
vector-builder-0.3.7.2: unregistering (dependencies changed)
vector-builder> configure (lib + test)
Configuring vector-builder-0.3.7.2...
vector-builder> build (lib + test)
Preprocessing library for vector-builder-0.3.7.2..
Building library for vector-builder-0.3.7.2..
[1 of 8] Compiling VectorBuilder.Prelude
[2 of 8] Compiling VectorBuilder.Core.Update
[3 of 8] Compiling VectorBuilder.Core.Builder
[4 of 8] Compiling VectorBuilder.MVector
[5 of 8] Compiling VectorBuilder.Builder
[6 of 8] Compiling VectorBuilder.Vector
[7 of 8] Compiling VectorBuilder.MonadPlus
[8 of 8] Compiling VectorBuilder.Alternative
Preprocessing test suite 'tests' for vector-builder-0.3.7.2..
Building test suite 'tests' for vector-builder-0.3.7.2..
[1 of 2] Compiling Main.Sample
[2 of 2] Compiling Main
Linking .stack-work/dist/x86_64-linux/Cabal-2.4.0.1/build/tests/tests ...
vector-builder> copy/register
Installing library in /home/simon/src/vector-builder/.stack-work/install/x86_64-linux/da6ab40722cf26def321f3afbad12b99af2f327f84d5d7d92f8138157548da6d/8.6.5/lib/x86_64-linux-ghc-8.6.5/vector-builder-0.3.7.2-Kobe1ABTFVWHtSalbGSjYI
Registering library for vector-builder-0.3.7.2..
vector-builder> Test running disabled by --no-run-tests flag.
Completed 2 action(s).
```

Now I modify the last module in build the order, and `stack` rebuilds from zero!

```
$ echo '-- bla' >> library/VectorBuilder/Alternative.hs 
$ stack build
vector-builder-0.3.7.2: unregistering (local file changes: library/VectorBuilder/Alternative.hs)
vector-builder> configure (lib)
Configuring vector-builder-0.3.7.2...
vector-builder> build (lib)
Preprocessing library for vector-builder-0.3.7.2..
Building library for vector-builder-0.3.7.2..
[1 of 8] Compiling VectorBuilder.Prelude
[2 of 8] Compiling VectorBuilder.Core.Update
[3 of 8] Compiling VectorBuilder.Core.Builder
[4 of 8] Compiling VectorBuilder.MVector
[5 of 8] Compiling VectorBuilder.Builder
[6 of 8] Compiling VectorBuilder.Vector
[7 of 8] Compiling VectorBuilder.MonadPlus
[8 of 8] Compiling VectorBuilder.Alternative
vector-builder> copy/register
Installing library in /home/simon/src/vector-builder/.stack-work/install/x86_64-linux/da6ab40722cf26def321f3afbad12b99af2f327f84d5d7d92f8138157548da6d/8.6.5/lib/x86_64-linux-ghc-8.6.5/vector-builder-0.3.7.2-JJTkYJEXEoEG9xl9pXropy
Registering library for vector-builder-0.3.7.2..
```

**Comment by mihaimaruseac** [comment]
Thanks. Will investigate when I get a few cycles for open source

**Comment by minad** [comment]
One thing to add - for me it builds things only once if I do `change something in the code; stack test; stack build` but twice if I do `change something in the code; stack build; stack test`.

**Comment by sjakobi** [comment]
> and I interrupt when the test suite is starting

You could use `--no-run-tests` for that.

**Comment by minad** [comment]
@sjakobi Ok that sounds good :see_no_evil: I have to admit I didn't really dig deep into all the things stack has to offer. I am using it for its great out of the box experience (Thank you stack developers for that!). And besides this problem I have not many other issues with it.

**Comment by mihaimaruseac** [comment]
Can also try `stack build --test --bench --no-run-tests --no-run-benchmarks` if you want to build everything but not run tests/benchmarks.

That's what I used to do at previous place where we were using Haskell daily

**Comment by rjmk** [comment]
Building with `stack test` appears to add the test dependencies to the dependencies of the library, thus invalidating the previous build

**Comment by jproyo** [comment]
Just in case if it helps i am facing the same issue with `2.1.3` but after doing this, some re linkings are gone:

`stack build && stack test components --test-arguments "-m MyComponent" --file-watch`

In the case i run with 

`stack test components --test-arguments "-m MyComponent" --file-watch`

It tries to relink everything on each change. With `stack build` on front it seems it is avoid some recompilation, even when i change only one space in the test.

**Comment by kindaro** [comment]
Is this a duplicate of #2800? Looks similar from the outset, but the explanation given is completely different _([wrong dependencies added to library][wrong_dependencies] here and [upstream issue][upstream_issue] there)_.

[wrong_dependencies]: https://github.com/commercialhaskell/stack/issues/4977#issuecomment-599858011
[upstream_issue]: https://github.com/commercialhaskell/stack/issues/2800#issuecomment-341658318

I am also seeing similar behaviour, and I am confused as to where to put a `+1`.

**Comment by wraithm** [comment]
@kindaro I would assume it's the same.

**Comment by natefaubion** [comment]
The [PureScript compiler](https://github.com/purescript/purescript) suffers from the same symptoms as reported here. I currently work around it by downgrading to Stack 1.9, which doesn't appear to have the same problems.

**Comment by anka-213** [comment]
I have this problem and it is completely consistent: every time I switch between `stack build` and `stack test` it recompiles everything.
I do have a custom `Setup.hs` which adds an extra pre-processor, but those files aren't regenerated, just recompiled so that shouldn't be the issue.

**Comment by qrilka** [comment]
switching between `build` and `test` invocations normally changes "source map" of dependencies/flags used. Component-based builds should improve the situation but that effort stalled. One possible workaround is to use different STACK_WORK values for `build` and `test`.

**Comment by jproyo** [comment]
> I have this problem and it is completely consistent: every time I switch between `stack build` and `stack test` it recompiles everything.
> I do have a custom `Setup.hs` which adds an extra pre-processor, but those files aren't regenerated when switching, just recompiled, so that shouldn't be the issue.

What is working for me pretty much lately is to use a different build command: 

```bash
stack build --test --no-run-tests
```

With this approach, you build all sources, including test, and after that, if you run `stack test`  it won't recompile.

The only problem with this is that if you are using some integrated environment with your IDE (haskell-server for example) this will launch a `stack build` behind the scenes and the previous will recompile if you IDE compiles you run `stack test`

**Comment by hasufell** [comment]
Try this in your `stack.yaml`:

```yml
build:
  test: true
  test-arguments:
    no-run-tests: true
  bench: true
  benchmark-opts:
    no-run-benchmarks: true
```

See https://docs.haskellstack.org/en/latest/yaml_configuration/#build

**Comment by anka-213** [comment]
@hasufell That sounds like a great solution! How do I run the tests then? Does `stack test` still work?

**Comment by Javran** [comment]
Now that I use `--no-run-tests` to avoid rebuilds, I wish there's a way to suppress  "Test running disabled" messages.

My current workaround is a mess, but at least it works, for easy copy-and-paste:

```
stack build --color=always --test --no-run-tests 2>&1 | grep --color=always -vP 'Test running disabled by --no-run-tests flag'
```

**Comment by mpilgrem** [comment]
I am new to this discussion. To summarise my own understanding:
* a 'test' is a package executable that is run automatically once it is built. You can stop that with flag `--no-run-tests`. Stack announces that it is not running the 'test' executable, as it normally would.
* `stack build` with no target specified does not build 'tests', unless flag `--test` is passed. `stack test` is a synonym for `stack build --test`.
* if you want to specify that a non-'test' executable should be run after it is built, you can make use of the `--exec` option (eg `stack build --exec my-executable`).
* if you build one executable with its dependencies and then ask to build a different executable with the same dependencies, Stack does not need to build some things again. So, it may be sensible to specify common dependencies for 'tests' and non-'tests', even if they are not actually used.
* if you build one executable with its dependencies and then ask to build a different executable with different depedencies, Stack (necessarily) rebuilds things.

**Comment by mpilgrem** [comment]
@Javran, Stack's announcement that `Test running disabled by --no-run-tests flag.` is at the level of 'information'. So, if you command `stack --verbosity warn test` or, equivalently, `stack  --verbosity warn build --test`, Stack will not provide that information. However, it will also not provide any other Stack message at the level of 'information'.

**Comment by mpilgrem** [comment]
I am going to close this issue given the passage of time and because I can't reproduce the original issue with Stack 2.15.5, namely:
~~~text
> stack new foo
> cd foo
> stack build
... builds lib + exe
> stack build --tests --no-run-tests
... builds test suite only
~~~

If it is still an issue with current versions, please reopen a fresh issue.

**Comment by wraithm** [comment]
omg @mpilgrem it's fixed. Holy cow. This is incredible. I can confirm  that this is working well.

**Comment by LeuschkeTressa** [comment]
Sharing my findings in case it is helpful to others. (Part of this is also conveyed in previous comments.)

Once `stack` has compiled all components (lib + exes + tests + benchmarks) _in one go_ (= a single invocation of `stack`), components can be re-built independently without causing unnecessary re-compilation.

So if we do...
```bash
stack clean
stack build --bench --no-run-benchmarks --test --no-run-tests
  # compiles everything
```

...then thereafter, as long as `stack` doesn't have an external reason to re-compile anything (such as a source file edit), the below command sequence will not cause any re-compilation (no sequence of the below commands will):

```bash
stack build --bench --no-run-benchmarks
stack build --test --no-run-tests
stack build
stack build --bench --no-run-benchmarks --test --no-run-tests
stack build --bench --no-run-benchmarks --test --no-run-tests :myComponent
# ...
```

Above I have added in `--no-run-benchmarks` and `--no-run-tests` so that test/bench components are compiled but not run - but actually runing these doesn't change anything.

If at any point `stack` re-compiles anything, e.g. because a source file has changed, then this "good state" is broken however - unless that re-compilation was a "compile all in one invocation"-compilation.

The "good state" can be brought back with:

```bash
# either: clean, then compile all in one go:
stack clean
stack build --bench --no-run-benchmarks --test --no-run-tests

# ...or: compile all in one go _twice_:
stack build --bench --no-run-benchmarks --test --no-run-tests
stack build --bench --no-run-benchmarks --test --no-run-tests
```

---

It seems that one way to always stay in the good state is to make sure that each invocation of `stack` that has the potential to re-compile "anything" is always a request to `stack` to "compile everything".

E.g. depending on what we actually want `stack` to do we can use invocations such as these:

```bash
# run all benchmarks:
stack build --bench                     --test --no-run-tests
    
# run all tests:
stack build --bench --no-run-benchmarks --test
    
# run only test suite component :myTest:
stack build --bench --no-run-benchmarks --test :myTest
```

Here's an example of a sequence of commands/actions + what `stack` will do:
```bash
# take us into the good state:
stack clean
stack build --bench --no-run-benchmarks --test --no-run-tests

# now we want to run all tests:
stack build --bench --no-run-benchmarks --test
  # (re-compiles nothing)
  # runs all tests

# HERE: edit a *test* source file

# now we want to run all benchmarks:
stack build --bench --test --no-run-tests
  # re-compiles the test component of the edited source file
  # (re-compiles nothing else)
  # some re-linking
  # runs all benchmarks

# now we want to run all tests:
stack build --bench --no-run-benchmarks --test
  # (compiles nothing)
  # runs all tests

```

---

Also possibly helpful in order to avoid unnecessary re-compilation is to make sure that all components have non-overlapping `source-dirs:` in `package.yaml` (for HPACK projects).
